cirq/value/linear_dict.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Linear combination represented as mapping of things to coefficients."""
+
+from typing import Any, Dict, Mapping, Union
+
+Scalar = Union[complex, float]
+
+
+class LinearDict(Dict[Any, Scalar]):
+    """Represents linear combination of things.
+
+    LinearDict implements the basic linear algebraic operations of vector
+    addition and scalar multiplication for linear combinations of abstract
+    vectors. Keys represent the vectors, values represent their coefficients.
+    The only requirement on the keys is that they be hashable.
+
+    A consequence of treating keys as opaque is that all relationships between
+    the keys other than equality are ignored. In other words, we treat linear
+    combinations as if all distinct keys are independent.
+
+    For convenience, methods which take a second linear combination accept
+    a standalone key instance as well. The second argument is then treated as
+    a single-term linear combination with the sole coefficient equal to 1.0.
+    """
+    def __init__(self, terms: Mapping[Any, Scalar]) -> None:
+        """Initializes linear combination from a collection of terms.
+
+        Args:
+            terms: Mapping of abstract vectors to coefficients in the linear
+                combination being initialized.
+        """
+        super(LinearDict, self).__init__()
+        self.update(terms)
+
+    def clean(self, *, tolerance: float=1e-9) -> None:
+        """Modifies self by deleting terms with small coefficients."""
+        negligible = [v for v, c in self.items() if abs(c) <= tolerance]
+        for v in negligible:
+            del self[v]
+
+    def copy(self) -> 'LinearDict':
+        return LinearDict(super(LinearDict, self).copy())
+
+    def __getitem__(self, vector: Any) -> Scalar:
+        if vector in self:
+            return super(LinearDict, self).__getitem__(vector)
+        return 0.0
+
```

Comment:
Override `__setitem__` so that it deletes when setting to 0.

Reply:
Initially agreed. Then I discovered this leads surprising behavior when writing loops such as:

```python
for vector in linear_combination:
    linear_combination[vector] *= something
```

The surprising behavior is the `RuntimeError` exception thrown when dictionary changes size during iteration. I think we should use inherited `__setitem__()` to avoid this. Users can call `clean()` to remove the zero terms. Presence of these terms should not have any consequences for observable behavior of `LinearDict`.

Reply:
We can bypass that issue by having `__iter__` return `list(super().__iter__())`

Reply:
Done. Employed this snapshot-grabbing approach in other similar places, too, e.g. `items()`.